### DISCUSSION TITLE
sendNoteOn(pitch, velocity, channel);     How to use??

### DISCUSSION THREAD

**Discussion by AaronHaCohen** [FortySevenEffects/arduino_midi_library#355] [Q&A]
Can anyone tell me how to use this function? I tried but it wouldn't compile in my otherwise error free Callbacks.ino code. The compiler complains that sendNoteOn() is undeclared. I thought that was handled in MIDI.hpp like the other functions in the Callbacks code. The documentation says "Definition at line [226] in MIDI.hpp. and curiously, line 230 says" "send(NoteOn, inNoteNumber, inVelocity, inChannel);
Is this a typo?  Thanks for any help, Aaron

Code snippet:
`  void handleNoteOn(byte channel, byte pitch, byte velocity) {
         sendNoteOn(pitch, velocity, channel);
}
 `

**Comment by lathoub** [comment]
Assuming the main object is MIDI, then: `MIDI.sendNoteOn(pitch, velocity, channel);`

**Reply by franky47** [reply]
@AaronHaCohen in C++, you can call functions on instances of a class/struct (we call those functions methods), the same way you would access a property/member of the struct in C: using the dot operator.

The `MIDI_CREATE_DEFAULT_INSTANCE` macro defines a global `MIDI` object, on which you can call methods defined in the docs (like `MIDI.begin`, `MIDI.turnThruOff` etc..).

Callbacks are a bit different: **you** define a function (plain old C function) to be called back when a message arrives, and to do so you _register_ it with `MIDI.setHandleXYZ`, by passing the name of your callback function (internally it's function pointers).

I hope this clears things out.

If you are trying to disable Thru and remap notes or messages, give the experimental (unreleased) [advanced Thru](https://github.com/FortySevenEffects/arduino_midi_library/issues/40#issuecomment-647012093) feature a try, it gives you finer control over filtering and remapping messages.

**Reply by AaronHaCohen** [reply]
Hello Franky,

Thanks for your advice and details. I may start to remember to use the dot operator soon, I hope. It's challenging for me to adapt to the Object Oriented methods and mentality after so many years of plain C procedural code with variables and functions.  I don't believe I ever used a structure in previous work. But we were writing close to the hardware level. We did use a union on occasion so there's still hope for me.

I'm starting to get the hang of it now. My code is now receiving NoteOns, making changes and sending them out. And likewise CC's in and different CC's out.
I'm still using Serial.write() to send stuff out and wonder if there is any advantage to using the class methods like sendNoteOn() etc. Faster? Less memory? More reliable?

Thanks also for the suggestion to try your advanced Thru feature. I'm interested to see how it might give finer control than just parsing and mapping bytes.

And I want to thank you for providing this whole FortySevenEffects library. It certainly makes receiving MIDI easy and reliable.

Thanks again, Aaron

**Reply by franky47** [reply]
> I'm still using Serial.write() to send stuff out and wonder if there is any advantage to using the class methods like sendNoteOn() etc. Faster? Less memory? More reliable?

None: it's just an abstraction providing a more user-friendly API, not having to know or care about how MIDI encodes the underlying messages.

**Comment by AaronHaCohen** [comment]
Hello Bart,

Thanks for the advice. Someday I might remember to use the dot operator. I
hope that day is soon. Thanks again, Aaron
